src/harness/sas_test_harness.py

```diff
@@ -0,0 +1,364 @@
+    self.server.socket = ssl.wrap_socket(
+        self.server.socket,
+        certfile=self.cert_file,
+        keyfile=self.key_file,
+        ca_certs=self.ca_cert_file,
+        cert_reqs=ssl.CERT_REQUIRED,
+        ssl_version=ssl.PROTOCOL_TLSv1_2,
+        ciphers=':'.join(CIPHERS),
+        server_side=True
+    )
+
+    # Start HTTP Server.
+    logging.info("Started Test Harness Server:%s at %s", self.name, self.http_server_url)
+    self.server.serve_forever()
+
+  def shutdown(self):
+    """ This method is used to stop HTTPServer Socket
+    """
+    self.server.shutdown()
+    logging.info("Stopped Test Harness Server:%s", self.name)
+
+  def cleanDumpFiles(self):
+    """ Clean existing dumpfile if any. """
+    temp_dir_path = self.getDumpFilePath()
+    map(os.remove, (os.path.join(temp_dir_path, file_path)
+                    for file_path in os.listdir(temp_dir_path)
+                    if os.path.exists(os.path.join(temp_dir_path, file_path))))
+    os.rmdir(temp_dir_path)
+    logging.info('all dump records generated by this testcase are cleaned successfully')
+    
+  def __createFadObject(self, all_activity_dump):
+    """Creates the FAD object format and embed the active dump file object of different recordtype.
+    """
+    full_activity_dump = {}
+    full_activity_dump['files'] = all_activity_dump
+    end_time = time.strftime(datetime.now().replace(microsecond=0).isoformat()) + 'Z'
+    full_activity_dump['generationDateTime'] = end_time
+    full_activity_dump['description'] = 'Load {} FAD to SAS UUT'.format(self.host_name)
+    return full_activity_dump
+
+  def __createFadRecord(self, encoded_url, record):
+    """ Creates the files object that includes url,checksum,size,version and recordtype of record
+        required to be packed in FAD.
+    """
+    fad_record = {}
+    fad_record['url'] = encoded_url
+    fad_record['checksum'] = hashlib.sha1(json.dumps(record)).hexdigest()
+    fad_record['size'] = sys.getsizeof(record)
+    fad_record['version'] = self.sas_version
+    if 'cbsd' in encoded_url:
+      fad_record['recordType'] = "cbsd"
+    elif 'zone' in encoded_url:
+      fad_record['recordType'] = "zone"
+    elif 'esc_sensor' in encoded_url:
+      fad_record['recordType'] = "esc_sensor"
+    else:
+      raise Exception("ConfigurationError:: Incorrect Record Type received in URL:{}".format(encoded_url))
+    return fad_record
+
+  def __writeDumpFile(self, file_name, data):
+    """ Write JSON data to specified file. """
+    dump_file_path = os.path.join(self.getDumpFilePath(), file_name)
+    with open(dump_file_path, "w") as file_handler:
+      file_handler.writelines(json.dumps(data, sort_keys=True, indent=4))
+
+  def writeFadRecords(self, dump_records_list):
+    """ This method is used to write dump records in specified path these files will be served
+        by SAS Test Harness.
+    Args:
+       dump_records_list: List of dump_records object required to be written on the dump path.
```

Comment:
This implementation isn't quite right. It took me awhile to understand, but it seems that `dump_records_list` is a list of CbsdData, ZoneData, and EscSensorData objects (all mixed together). From the 'id' field which happens to exist in each of these objects, you auto-determine the type of record, pick a filename and URL, then write the each record to its own file. Then you assemble the metadata into FullActivityDump object which is written to "FAD.json".

This does not provide the necessary flexibility to specify e.g. 2 CBSDs in one file and 10 in another file. Although I agree that it should not make a difference to the SAS UUT, ITS needs the capability to verify this.

I believe that the right format for the input is a list of objects, each with the following fields:
 - Filename (or relative path?) -- this will also be used in constructing the URL
 - "body": List of e.g. CbsdData objects (all objects must be of the same type)

Then the function will:
 - Write the body directly to the specified file.
 - Construct the URL using the base path of the SAS TH and the filename given.
 - Append the information to the FAD record.

Reply:
to be addressed in a separate submission. We are working on it right now.

Reply:
We have address this issue in the latest submission. Updated code to dump multiple records of the same type in the same file. Also updated logic to create multiple files for cbsd records - this is to demonstrate the capability that when multiple dump file of each record type is needed in case that, for example, too many of them that cannot be packed into one dump file. For example cbsd1-cbsd5 in activity_dump_file_cbsd0.json file and cbsda-cbsdc in activity_dump_file_cbsd1.json.